# Pipeline for processing data from 10X genomics [Single_cell_basic_processing_rev4.nf](../../nextflow/Single_cell_basic_processing_rev4.nf)
This pipeline performs basic analysis of data generated using the 10X genomics protocol.  
It produces gene x cell count matrices using the [Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) software. 
In addition an RNA velocity analysis is run using [velocyto](velocyto.org/velocyto.py/index.html) and ready to use [Scanpy](https://scanpy.readthedocs.io) AnnData objects are created.
**Note**: The Nextflow script of the pipeline has been automatically generated by [DolphinNext](https://dolphinnext.readthedocs.io/en/latest/).

# Software used by this pipeline and install instructions
* R (Rscript) can be installed using conda [https://anaconda.org/r/r](https://anaconda.org/r/r)
* [Nextflow](https://www.nextflow.io/) can be installed using conda [https://anaconda.org/bioconda/nextflow](https://anaconda.org/bioconda/nextflow)
* Samtools can be installed using conda [https://anaconda.org/bioconda/samtools](https://anaconda.org/bioconda/samtools)
* [Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) (**Note:** This pipline has been used and tested with cellranger-2.1.1)
* Python 3 and packages numpy, pandas and scanpy (can be installed using pip)
  
Make sure all tools can be found, edit your $PATH if necessary.  

# Input: samplefile.tsv

Create a tab-separated file containing the samples. The sample header must consist of the following mandatory fields.  

* Sample: Name of the sample (must match the names of the fastq files)
* ncells: Estimated number of cells
* fastqpath: Path to directory containing the sequencing reads in fastq format
* refindex: Path to the reference genome index (see this [link](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/references) on how to create custom references)
* refgtf: Path to gtf file describing the gene models (is available in the genome index folder created by Cell Ranger in genes/genes.gtf)
* chemistry: Version of the chemistry (e.g. SC3Pv2, see this [link](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count) for more info.)

# Output folders
cellranger/$SAMPLENAME: Cellranger output  
samplefiles/$SAMPLENAME: Sample parameter files  
scanpy_AnnData/$SAMPLENAME/$genome: Scanpy AnnData h5ad files  
velocyto/$SAMPLENAME/$genome: Velocyto loom files  

# How to run
`nextflow Single_cell_basic_processing_rev4.nf --samplesheet samplefile.tsv --ncpus 4 --maxJobs 200 -with-report -with-trace -with-timeline -with-dag`

**Parameters:**  
The CellRanger pipeline is configured to run in the SGE mode and use a cluster. To run in this mode you need an additional nextflow.config file with the cluster configuration (queue name, resources etc., see [nextflow documentation](https://www.nextflow.io/docs/latest/index.html) for instructions).  
The --maxJobs parameter controls how many jobs are at max submitted to the cluster.  
If you want to run the pipeline on a single host, please adapt the cellranger command in the nextflow script accordingly (for info on options use `cellranger count --help`).  
The Parameter "--ncpus" defines the number of cores to be used by samtools and velocyto.  
